test/ERC20LockUpScenarios.test.ts

```diff
@@ -0,0 +1,328 @@
+    let rewardTokenPerSecond: BigNumberish;
+    let poolStartTime: BigNumberish;
+    let poolEndTime: BigNumberish;
+    let unstakeLockUp: BigNumberish;
+    let claimLockUp: BigNumberish;
+    let ipfsHash: BytesLike;
+    let signer: HardhatEthersSigner;
+    let ayo: HardhatEthersSigner;
+    let alina: HardhatEthersSigner;
+    let vartan: HardhatEthersSigner;
+    let nikita: HardhatEthersSigner;
+    let mary: HardhatEthersSigner;
+    let mansur: HardhatEthersSigner;
+  
+  
+    beforeEach(async function () {
+  
+      // Get signers
+      [signer, ayo, alina, vartan, signer, nikita, mary, mansur] = await ethers.getSigners();
+  
+      mockStakeToken = await ethers.deployContract("ERC20MockToken", [
+        "StakeToken",
+        "STK",
+        18
+      ]);
+  
+      mockRewardToken = await ethers.deployContract("ERC20MockToken", [
+        "RewardToken",
+        "RTK",
+        18
+      ]);
+      //First mint reward tokens for user before activating pool
+      await mockStakeToken.mint(
+        signer.address,
+        ethers.parseEther("2000000000")
+      );
+      await mockRewardToken.mint(
+        ayo.address,
+        ethers.parseEther("2000000000")
+      );
+  
+      ipfsHash = ethers.randomBytes(32);
+      rewardTokenPerSecond = ethers.parseEther("1");
+      poolStartTime = (await time.latest()) + 100; // Start in 100 seconds
+      poolEndTime = poolStartTime + 1000; // End in 1000 seconds after start
+      unstakeLockUp = poolStartTime + 500; // Lockup 500 seconds after start
+      claimLockUp = poolStartTime + 200; // Lockup 200 seconds after start
+      rewardTokenPerSecond = ethers.parseEther("1"); // 1 RWD per second
+    });
+  
+    it("Handles standard staking, unstaking, claiming, and pool ending", async function () {
+      // Set pool parameters
+      poolStartTime = (await time.latest()) + 100; // Start in 100 seconds
+      poolEndTime = poolStartTime + 1000; // End in 1000 seconds after start
+      unstakeLockUp = poolStartTime + 500; // Lockup 500 seconds after start
+      claimLockUp = poolStartTime + 200; // Lockup 200 seconds after start
+      rewardTokenPerSecond = ethers.parseEther("1"); // 1 RWD per second
+      const data = {
+        stakeToken: await mockStakeToken.getAddress(),
+        rewardToken: await mockRewardToken.getAddress(),
+        poolStartTime: poolStartTime,
+        poolEndTime: poolEndTime,
+        unstakeLockUpTime: unstakeLockUp,
+        claimLockUpTime: claimLockUp,
+        rewardPerSecond: rewardTokenPerSecond
+      }
+      let { poolContract } = await deployAndSetupPool(ipfsHash, ayo, data)
+      // **Scenario: Pool Not Started**
+      await expect(poolContract.stake(ethers.parseEther("100")))
+        .to.be.revertedWithCustomError(poolContract, "PoolNotStarted");
+  
+      // **Scenario: Pool Started, Staking**
+      await time.increase(100); // Advance time to pool start
+      await mint(100, alina, mockStakeToken)
```

Comment:
Do not mint tokens to users, mint them only once and then transfer to other users